core/federated/network/socket_common.c

```diff
@@ -0,0 +1,410 @@
+
+  int count = 1;
+  while (result != 0 && count++ < PORT_BIND_RETRY_LIMIT) {
+    if (specified_port == 1) { // RTI is set to 1 if no specified port.
+      lf_print_warning("RTI failed to get port %d.", used_port);
+      used_port++;
+      if (used_port >= DEFAULT_PORT + MAX_NUM_PORT_ADDRESSES)
+        used_port = DEFAULT_PORT;
+      lf_print_warning("RTI will try again with port %d.", used_port);
+      server_fd.sin_port = htons(used_port);
+      // Do not sleep.
+    } else {
+      lf_print("Failed to bind socket on port %d. Will try again.", used_port);
+      lf_sleep(PORT_BIND_RETRY_INTERVAL);
+    }
+    result = bind(socket_descriptor, (struct sockaddr*)&server_fd, sizeof(server_fd));
+  }
+
+  // Set the global server port.
+  if (specified_port == 0) { // Federates are set to 0 if no specified port.
+    // Need to retrieve the port number assigned by the OS.
+    struct sockaddr_in assigned;
+    socklen_t addr_len = sizeof(assigned);
+    if (getsockname(socket_descriptor, (struct sockaddr*)&assigned, &addr_len) < 0) {
+      lf_print_error_and_exit("Failed to retrieve assigned port number.");
+    }
+    used_port = ntohs(assigned.sin_port);
+  }
+  if (result != 0) {
+    lf_print_error_and_exit("Failed to bind the RTI socket. Port %d is not available. ", used_port);
+  }
+  return used_port;
+}
+
+void create_server(uint16_t port, int* final_socket, uint16_t* final_port, int sock_type) {
+  int socket_descriptor;
+  struct timeval timeout_time;
+  if (sock_type == 0) {
+    // Create an IPv4 socket for TCP.
+    socket_descriptor = create_real_time_tcp_socket_errexit();
+    // Set the timeout time for the communications of the server
+    timeout_time =
+        (struct timeval){.tv_sec = TCP_TIMEOUT_TIME / BILLION, .tv_usec = (TCP_TIMEOUT_TIME % BILLION) / 1000};
+  } else {
+    // Create a UDP socket.
+    socket_descriptor = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+    timeout_time =
+        (struct timeval){.tv_sec = UDP_TIMEOUT_TIME / BILLION, .tv_usec = (UDP_TIMEOUT_TIME % BILLION) / 1000};
+  }
+  char* type = (sock_type == 0) ? "TCP" : "UDP";
+  if (socket_descriptor < 0) {
+    lf_print_error_system_failure("Failed to create %s socket.", type);
+  }
+  set_socket_timeout_option(socket_descriptor, &timeout_time);
+  int used_port = set_socket_bind_option(socket_descriptor, port);
+  if (sock_type == 0) {
+    // Enable listening for socket connections.
+    // The second argument is the maximum number of queued socket requests,
+    // which according to the Mac man page is limited to 128.
+    listen(socket_descriptor, 128);
+  }
+  *final_socket = socket_descriptor;
+  *final_port = used_port;
+}
+
+void create_TCP_server(uint16_t port, int* final_socket, uint16_t* final_port) {
+  create_server(port, final_socket, final_port, 0);
+}
+void create_UDP_server(uint16_t port, int* final_socket, uint16_t* final_port) {
+  create_server(port, final_socket, final_port, 1);
+}
```

Comment:
The `create_UDP_server()` function is currently only used for creating the clock synchronization server.
<pre>
  if (rti_remote->clock_sync_global_status >= clock_sync_on) {
    <b>create_UDP_server(rti_remote->final_port_TCP + 1, &rti_remote->socket_descriptor_UDP, &rti_remote->final_port_UDP);</b>
  }
  return rti_remote->socket_descriptor_TCP;
}
</pre>


I split these functions into two, because, I wanted to change that part's usage as `create_clock_sync_server()` when I integrate the network driver.

I think there is no reason for the clock synchronization be done by anything other than UDP. 
If we have no plans except using UDP, I can add this function in `clock-sync.c`.

So, is it worth to use other network stacks for the clock-synchronization?

Reply:
I think clock sync really needs to use UDP, but there are currently no plans to use it for anything else.

Reply:
Then would it be fine to leave this as is? I would not like to leave the `socket_type_t` as an argument, for further development.

Reply:
@edwardalee I removed the separate functions, and just left `create_server()`. It is in this commit https://github.com/lf-lang/reactor-c/pull/505/commits/b2a5af8206f7a973d04eba8661d8978889fd2ed9.